core/src/apps/management/recovery_device/layout.py

```diff
@@ -68,24 +73,76 @@
             word = await ctx.wait(keyboard)
 
         if mnemonic_type == mnemonic.TYPE_SLIP39 and mnemonics:
-            # check if first 3 words of mnemonic match
-            # we can check against the first one, others were checked already
-            if i < 3:
-                share_list = mnemonics[0].split(" ")
-                if share_list[i] != word:
-                    raise IdentifierMismatchError()
-            elif i == 3:
-                for share in mnemonics:
-                    share_list = share.split(" ")
-                    # check if the fourth word is different from previous shares
-                    if share_list[i] == word:
-                        raise ShareAlreadyAddedError()
+            if not advanced_shamir:
+                # check if first 3 words of mnemonic match
+                # we can check against the first one, others were checked already
+                if i < 3:
+                    share_list = mnemonics[0].split(" ")
+                    if share_list[i] != word:
+                        raise IdentifierMismatchError()
+                elif i == 3:
+                    for share in mnemonics:
+                        share_list = share.split(" ")
+                        # check if the fourth word is different from previous shares
+                        if share_list[i] == word:
+                            raise ShareAlreadyAddedError()
+            else:
+                # in case of advanced shamir recovery we only check 2 words
+                if i < 2:
+                    share_list = mnemonics[0].split(" ")
+                    if share_list[i] != word:
+                        raise IdentifierMismatchError()
```

Comment:
We can check the first and second word every time (before the if branch). And then only for non-advanced Shamir check word 3 and 4.

Reply:
I think we will also need to check the 3rd word within the groups otherwise the user will have to enter the whole share to learn he already entered it.

Reply:
but we'll probably leave that for refactor in next milestone